components/compliance-service/reporting/relaxting/suggestions.go

```diff
@@ -502,6 +502,167 @@ func (backend ES2Backend) getControlSuggestions(client *elastic.Client, typePara
+	if typeParam == "control_tag_value" && controlTagFilterKey == "" {
+		return nil, errors.New("'control_tag' filter is required for 'control_tag_value' suggestions")
+	}
+	esIndex, err := GetEsIndex(filters, useSummaryIndex, true)
+	if err != nil {
+		return nil, errors.Wrap(err, "getControlTagsSuggestions unable to get index dates")
+	}
+	boolQuery := backend.getFiltersQuery(filters, true)
+	text = strings.ToLower(text)
+
+	var finalInnerQuery elastic.Query
+	if len(text) >= 2 {
+		finalInnerBoolQuery := elastic.NewBoolQuery()
+		finalInnerBoolQuery.Must(elastic.NewMatchQuery(fmt.Sprintf("%s.engram", target), text).Operator("or"))
+		finalInnerBoolQuery.Should(elastic.NewMatchQuery(fmt.Sprintf("%s.engram", target), text).Operator("and"))
+		finalInnerBoolQuery.Should(elastic.NewTermQuery(fmt.Sprintf("%s.lower", target), text).Boost(100))
+		finalInnerBoolQuery.Should(elastic.NewPrefixQuery(fmt.Sprintf("%s.lower", target), text).Boost(100))
+		finalInnerQuery = finalInnerBoolQuery
+	} else {
+		finalInnerQuery = elastic.NewExistsQuery(target)
+	}
+
+	outerBoolQuery := elastic.NewBoolQuery()
+	nestedBoolQuery := outerBoolQuery.Must(elastic.NewNestedQuery("profiles.controls.string_tags", finalInnerQuery))
+
+	hit := elastic.NewInnerHit().Size(20)
+	boolQuery = boolQuery.Must(elastic.NewNestedQuery("profiles.controls", nestedBoolQuery).InnerHit(hit))
+
+	searchSource := elastic.NewSearchSource().
+		Query(boolQuery).
+		FetchSource(false).
+		Size(size * 50) // Multiplying size to ensure that same profile with multiple versions is not limiting our suggestions to a lower number
+		// ^ Because we can't sort by max_score of the inner hits: https://discuss.elastic.co/t/nested-objects-hits-inner-hits-and-sorting/32565
+
+	source, err := searchSource.Source()
+	if err != nil {
+		return nil, errors.Wrap(err, "getControlTagsSuggestions unable to get Source")
+	}
+
+	LogQueryPartMin(esIndex, source, "getControlTagsSuggestions query")
+
+	searchResult, err := client.Search().
+		SearchSource(searchSource).
+		Index(esIndex).
+		FilterPath(
+			"took",
+			"hits.total",
+			"hits.hits._id",
+			"hits.hits._score",
+			"hits.hits.inner_hits").
+		Do(context.Background())
+
+	if err != nil {
+		logrus.Error("getControlTagsSuggestions search failed")
+		return nil, err
+	}
+
+	logrus.Debugf("Search query took %d milliseconds\n", searchResult.TookInMillis)
+
+	type ControlSourceTags struct {
+		ID         string                            `json:"id"`
+		StringTags []ESInSpecReportControlStringTags `json:"string_tags"`
+	}
+	type ControlTagsScore struct {
+		Score      float32
+		StringTags []ESInSpecReportControlStringTags
+	}
+
+	// Using this array to aggregate the unsorted suggestions from the response
+	foundControlTags := make([]ControlTagsScore, 0)
+	if searchResult != nil {
+		for _, hit := range searchResult.Hits.Hits {
+			if hit != nil {
```

Comment:
Instead of wrapping the body of each `for` loop in an `if`, you could probably remove several levels of nesting here by doing something like:

```
if hit == nil {
  continue
}
```

Reply:
I followed your advice Pete and removed 3 layers of nesting in another commit to this PR. Cheers!